design/volume-snapshot-data-movement/volume-snapshot-data-movement.md

```diff
@@ -663,6 +663,196 @@ If DM restarts, Velero has no way to detect this, DM is expected to:
 
 At present, VBDM doesn't support recovery, so it will follow the second rule.  
 
+## Kopia For Block Device
+To work with block devices, VGDP will be updated. Today, when Kopia attempts to create a snapshot of the block device, it will error because kopia does not support this file type. Kopia does have a nice set of interfaces that are able to be extended though.
+
+To achieve the necessary information to determine the type of volume that is being used, we will need to pass in the volume mode in provider interface.
+
+```go
+type PersistentVolumeMode string 
+
+const (
+	// PersistentVolumeBlock means the volume will not be formatted with a filesystem and will remain a raw block device.
+	PersistentVolumeBlock PersistentVolumeMode = "Block"
+	// PersistentVolumeFilesystem means the volume will be or is formatted with a filesystem.
+	PersistentVolumeFilesystem PersistentVolumeMode = "Filesystem"
+)
+
+// Provider which is designed for one pod volume to do the backup or restore
+type Provider interface {
+	// RunBackup which will do backup for one specific volume and return snapshotID, isSnapshotEmpty, error
+	// updater is used for updating backup progress which implement by third-party
+	RunBackup(
+		ctx context.Context,
+		path string,
+		volMode PersistentVolumeMode,
+		realSource string,
+		tags map[string]string,
+		forceFull bool,
+		parentSnapshot string,
+		updater uploader.ProgressUpdater) (string, bool, error)
+```
+
+In this case, we will extend the default kopia uploader to add the ability, when a given volume is for a block mode and is mapped as a device, we will use a new specific type that conforms to the [File](https://pkg.go.dev/github.com/kopia/kopia/fs@v0.13.0#File) interface. This new type, will handle streaming the block device to the rest of the kopia process.
+
+For this particular one, we will create a new `deviceEntry` and `deviceReader` structs that conform to the [fs.Reader](https://pkg.go.dev/github.com/kopia/kopia/fs@v0.13.0#Reader) and [File](https://pkg.go.dev/github.com/kopia/kopia/fs@v0.13.0#File) interfaces. 
+
+```go
+type deviceEntry struct {
```

Comment:
I didn't try it, but I am strongly feeling that the current requirement done by hacking `fs.File` and `fs.Reader` could be fulfilled by Kopia upstream's virtualfs directly. Specifically:

- It implements `fs.Entry` by `virtualEntry` which could replace `deviceEntry`
- It implements a `fs.StreamingFile` which is used similarly to `fs.File`, but it additionally allows the caller to provider an `io.ReadCloser`. Then I think it could replace `deviceReader`. Though it doesn't expose `io.Seek`, I think we don't need it, because with the current approach of block level access, we read the blocks sequentially from beginning to end all the time, we don't need to skip any blocks

If we can make this, we will have the Kopia code integration more predictably, because we know the design purpose for the Kopia upstream's StreamingFile, it won't break us in future if we follow its design purpose.
While for `fs.File` and `fs.Reader`, they are for file system level backups,  we don't know how they are going to develop in future version of Kopia.

Moreover, this virtualfs could help us to solve any potential existing/future conflicts that saving "not-really-file-system-data" into a file system data originated repo.